### PR TITLE
Fix DNSExit provider when provided with a zone and a non-identical hostname

### LINKED ISSUE #673: DNSExit: Updater breaks when provided with a zone and a non-identical hostname (bug)
### Description
When a configuration is provided with a zone and a hostname that is not identical to the zone, the updater fails. This issue was introduced with the implementation of the dnsexit provider, due to a misunderstanding of how the dnsexit API expects updates for 'subdomains'. This bug has been confirmed via the report here: [opnsense/plugins](https://github.com/opnsense/plugins/issues/3712)

### Steps to reproduce

1. Provide a configuration with a zone and a hostname that is not identical to the zone.
2. Run the updater.

### Expected behavior
The updater should produce the following JSON call for host `host1.myroot.publicvm.com` on zone `myroot.publicvm.com`:

```json
{
"domain":"myroot.publicvm.com",
"apikey":"myApiKey"
"update":
  [{
    "ttl":5,
    "type":"A",
    "content":"1.2.3.4",
     "name":"host1"
   }]
}
```

### Actual behavior
Currently, the name field is identical to the full hostname, which does not work.
Note that for calls with identical `name` and `zone`, this does work, as does an empty `name` with a `zone` that contains a root hostname (i.e. myroot.publicvm.com)

### Proposed solution
The issue can be fixed by:

- Only providing the 'hostname' part on the name field in the update block, by trimming it from the zone (if provided).
- Ensuring that this fix does not break other use cases (simple hostname only, without zone, hostname and zone being equal). Also handling ipv4 and ipv6 which must be done in a single call.

I will propose a solution shortly.

### PR DESCRIPTION
**Pull request by jortkoopmans**
This PR proposes a fix for issue #673. The fix involves trimming the zone from the hostname in the request. In addition to the bug fix, this PR includes refactoring for readability and the addition of a testing harness.

- Trimmed the zone from the hostname in the request to fix issue.
- Refactored code for improved readability.
- Added a testing harness which includes:
  - Moving JSON support loading into the sub for testing.
  - Creating a config loader helper sub.

### Testing
The actual bugfix is fairly limited, however, I ended up refactoring quite a lot to be able to write some tests. The current architecture of the application made it challenging to test the provider implementation due to:

- The %config hash being modified globally within the ddclient script (by several subs).
- The functions (subs) in ddclient not being exported as part of a module, making it difficult to use these functions in tests (outside of the script).

Despite these challenges, I've included some tests in this PR. I understand that this may trigger a discussion about how to improve the structure of the software, which I hope will be to the benefit of this program altogether.

fixes #673

### CODE REVIEW

File: ddclient.in [hunk 1]
```diff
@@ -4158,154 +4164,208 @@ EoEXAMPLE
...
-            info("Going to update IPv$ipv address to %s for %s.", $ip, $h);
-            $config{$h}{'status-ipv$ipv'} = 'failed';
-            $total_payload{$ipv} = {
-                name => $h,
-                type => $type,
-                content => $ip,
-                ttl => $config{$h}{'ttl'},
-            };
+            # One key per ipv (4 or 6)
+            my ($name, $modified_config) = $set_name_and_check_zone->($hostname, $config);
```
Comment (rhansen) [inline ddclient.in:4246]: The helper functions and config copy seem unnecessary to me.  Wouldn't something more direct like the following work?

```perl
            my $zone = $config->{'zone'} || $hostname;
            my $name = $hostname;
            if ($name !~ s/(?:^|\.)\Q$zone\E$//) {
                failed("Hostname %s does not end with the zone %s, cannot be trimmed", $hostname, $zone);
                next;
            }
```

**Review by rhansen — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: ddclient.in [hunk 2]
```diff
@@ -1823,6 +1823,11 @@ sub _read_config {
 
     return $content;
 }
+# Function to set %config for testing
+sub set_test_config {
+    my ($test_config) = @_;
+    %config = %$test_config;
+}
```
[the remark this answers is not in the document]
Comment (jortkoopmans) [reply ddclient.in]: Excellent point :+1: . Adopted.

File: t/dnsexit2.pl [hunk 3]
```diff
@@ -0,0 +1,222 @@
+use Test::More;
+use JSON::PP qw(encode_json decode_json);
+SKIP: { eval { require Test::Warnings; } or skip($@, 1); }
+eval { require 'ddclient'; } or BAIL_OUT($@);
+# To aid in debugging, uncomment the following lines. (They are normally left commented to avoid
+# accidentally interfering with the Test Anything Protocol messages written by Test::More.)
+# STDOUT->autoflush(1);
+# $ddclient::globals{'debug'} = 1;
+
+# Mock geturl to capture the parameters
```
[the remark this answers is not in the document]
Comment (jortkoopmans) [reply t/dnsexit2.pl]: Ok, I've implemented this, but it was a bit trickier to get the requested JSON out, as you can't seem to be able to share variables between these processes. I've use the UserAgent for this, but I had to also modify the CI a bit for this.

File: ddclient.in [hunk 4]
```diff
@@ -4158,154 +4164,208 @@ EoEXAMPLE
...
+    '5' => [ 'error',   'JSON Defined Record Type not Supported. Your JSON may try to update some record type not supported by our system.' ],
+    '6' => [ 'error',   'System Error. Our system problem. May not be your problem. Contact our support if you got such error.' ],
+    '7' => [ 'error',   'Error getting post data. Our server has problem to receive your JSON posting.' ],
+};
+
 sub nic_dnsexit2_update {
     debug("\nnic_dnsexit2_update -------------------");
+    
+    # Load json support directly, without init_config messing up the test config
+    load_json_support();
```
[the remark this answers is not in the document]
Comment (jortkoopmans) [reply ddclient.in]: Certainly, the main challenge is the ordering of the sub's, as I prefer only to test the dnsexit2 provider. However, the JSON support is loaded in the init_config sub. Now I tried calling init_config as well, but then it obviously modifies the configuration and I both need to then prime the initial configuration for the init_config to work, as well as later on override it again with the test values. This seemed very tedious in comparison to loading it myself in the provider sub.

File: ddclient.in [hunk 5]
```diff
@@ -4158,154 +4164,208 @@ EoEXAMPLE
...
+    # Local subroutine to verify if the zone is defined
+    my $check_zone = sub {
+        my ($config) = @_;
+        if ( not defined $config->{'zone'} or $config->{'zone'} eq ''){
+            return 1;
+        }
+        return 0;
+    };
```
[the remark this answers is not in the document]
Comment (jortkoopmans) [reply ddclient.in]: Good point, I've refactored to get rid of some of the heavy subs. Hope it's better this way.

File: ddclient.in [hunk 6]
```diff
@@ -4158,154 +4164,208 @@ EoEXAMPLE
...
-             '7'        => [ 'error',   'Error getting post data. Our server has problem to receive your JSON posting.' ],
-        );
-        my $ipv4 = delete $config{$h}{'wantipv4'};
-        my $ipv6 = delete $config{$h}{'wantipv6'};
+    # Local subroutine to set the name and check the zone
+    my $set_name_and_check_zone = sub {
+        my ($hostname, $config) = @_;
+
+        # Create a copy of the $config hash
+        my %config_copy = %$config;
```
[the remark this answers is not in the document]
Comment (jortkoopmans) [reply ddclient.in]: Fixed as well, indeed not needed.

File: ddclient.in [hunk 7]
```diff
@@ -4158,154 +4164,208 @@ EoEXAMPLE
...
+            return ('', \%config_copy);
+        }
+
+        # If $config_copy{zone} is set, create $name by trimming the zone from the hostname.
+        my $name = $hostname;
+        my $zone = $config_copy{'zone'};
+        if ($hostname =~ /\Q$zone\E$/) {
+            $name =~ s/\.?\Q$zone\E$//;
+        } else {
+            failed("Hostname %s does not end with the zone %s, cannot be trimmed", $hostname, $zone);
```
[the remark this answers is not in the document]
Comment (jortkoopmans) [reply ddclient.in]: Added a next;

File: ddclient.in [hunk 8]
```diff
@@ -4158,154 +4164,208 @@ EoEXAMPLE
...
+    my $set_name_and_check_zone = sub {
+        my ($hostname, $config) = @_;
+
+        # Create a copy of the $config hash
+        my %config_copy = %$config;
+
+        # If $config_copy{zone} is empty (''), return $name = ''
+        if ($check_zone->(\%config_copy)){
+            debug("Zone not defined or empty for hostname: %s", $hostname);
+            return ('', \%config_copy);
```
[the remark this answers is not in the document]
Comment (jortkoopmans) [reply ddclient.in]: Both would be fine according to the provider api docs, but I see why returning the full $hostname may sound more intuitive. I have modified it to do so.

File: ddclient.in (hunk 1 above)
Comment (jortkoopmans) [reply ddclient.in]: Yes, I've made this shorter / more to the point.

**Comment by jortkoopmans** [comment]
Amazing work @rhansen :star_struck: , it's looking great!
Good solution to revert to empty name by smart trimming, that simplifies the approach. And generally more concise coding :smiley: .
I have added a minor improvement, but more importantly I've tested the solution on the DNSExit API and it works fine.

**Review by rhansen — APPROVED** [review]
@jortkoopmans Thanks for reviewing.  I squashed all of the commits together then split into four:
  1. Add tests (marked as expected to fail)
  2. Fix bug (and mark tests as expected to pass)
  3. Don't skip remaining hosts on HTTP error.  This commit exists mostly to simplify the next commit: a `next` can be simply converted into a `return` whereas `last` needs additional logic in the calling function.
  4. Move the body of the `for` loop to a separate function

I split this PR into multiple commits for the following reasons:
  * It makes it easier for downstreams to cherry-pick a minimal diff to fix the issue.
  * It makes it easier for future developers to see what changed.
  * `git bisect` and `git revert` will be more useful if this PR breaks something.

Please review the commits one last time (especially the commit messages), make any changes you deem appropriate (force push is OK), and let me know if everything looks OK.

**Comment by jortkoopmans** [comment]
LGTM @rhansen . I have reviewed the commits, commit messages and re-tested the final solution just to be sure; all is good.

**Pull request merged**